homeassistant/components/nut/__init__.py

```diff
@@ -1 +1,159 @@
+    """Set up the Network UPS Tools (NUT) component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Network UPS Tools (NUT) from a config entry."""
+
+    config = entry.data
+    host = config[CONF_HOST]
+    port = config[CONF_PORT]
+
+    alias = config.get(CONF_ALIAS)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+
+    data = PyNUTData(host, port, alias, username, password)
+
+    status = await hass.async_add_executor_job(pynutdata_status, data)
+
+    if not status:
+        _LOGGER.error("NUT Sensor has no data, unable to set up")
+        raise ConfigEntryNotReady
+
+    _LOGGER.debug("NUT Sensors Available: %s", status)
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        PYNUT_DATA: data,
+        PYNUT_STATUS: status,
+        PYNUT_UNIQUE_ID: unique_id_from_status(status, host, port),
+    }
+
+    entry.add_update_listener(_async_update_listener)
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+def unique_id_from_status(status, host, port):
+    """Find the best unique id from the status."""
+    # The serial number should always exist, however some off brand
+    # UPSs do not send one so we use the next best thing
+    return status.get("device.serial") or status.get("ups.serial") or f"{host}:{port}"
+
+
+def find_resources_in_config_entry(config_entry):
+    """Find the configured resources in the config entry."""
+    if CONF_RESOURCES in config_entry.options:
+        return config_entry.options[CONF_RESOURCES]
+    return config_entry.data[CONF_RESOURCES]
+
+
+def pynutdata_status(data):
+    """Wrap for data update as a callable."""
+    return data.status
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class PyNUTData:
+    """Stores the data retrieved from NUT.
+
+    For each entity to use, acts as the single point responsible for fetching
+    updates from the server.
+    """
+
+    def __init__(self, host, port, alias, username, password):
+        """Initialize the data object."""
+
+        self._host = host
+        self._port = port
+        self._alias = alias
+        self._username = username
+        self._password = password
+
+        self.pynuterror = PyNUTError
```

Comment:
We don't need to store the error as an instance attribute. That's legacy from when we couldn't import names at the top.

Reply:
I was wondering about that ( I didn't change anything in this class to keep scope down ).  Will remove.  Thank you.